audbackend/core/backend/base.py

```diff
@@ -69,12 +69,12 @@ def _assert_equal_checksum(
 
         """
         if path_is_local:
-            checksum = utils.checksum(path)
+            checksum = audeer.md5(path)
```

Comment:
At first, it felt a little weird seing the identifiers `md5sum` and `checksum` so close together feels.  MD5  is already a type of checksum.  

 `audbackend.checksum` calls the private method `_checkum` of the concrete backend implementation . For example in minio `stat_object.metadata` of the client is fetched, and the checksum is extracted from the "x-amz-meta-checksum". For Artifactory the method is called md5 from the `.stat` field of a path. So checksum can be an md5 but does not have to be because it uses a backend implementation. 

Would it make sense to factor the checksum call into a  private method for pure readability that contains a docstring that contains this information (cross-referencing the backend implementations)?

Reply:
I also find that `_assert_equal_checksum`:
path represents the current file being processed or checked, and 
path_ref represents a reference file against which the first file's checksum is being compared.

I find the information that the function makes no assumption about the `path` and ` path_ref` valuable. In other words both could be local or both could be remote files.  Despite being a private method, would a docstring be a possibility to clarify this?

Reply:
Is the diagram in https://github.com/audeering/audbackend/pull/255#issuecomment-2493135995 correct?
My understanding is that `backend.put_file` triggers the file upload. There is a class `Interface` but it is not relevant for the backend interaction.

Reply:
The former introduced function `audbackend.checksum()` was implemented in `audbackend.core.utils.checksum()`. It was used on files stored at the current local disc of a user (basically, files outside of the backend the files should be uploaded to). This is given by the `if path_is_local` command above. Those cases are now replaced by calling `audeer.md5()` as we did before https://github.com/audeering/audbackend/pull/245.
This also means `audbackend.checksum()` does not call the private `_checkum()` method. This method is instead called by all backend methods deriving from `audbackend.backend.Base.checksum()`, and are meant to get the checksum from a backend (Artifactory, S3, filesystem). For the filesystem backend, it also calls `audeer.md5()`.

> Would it make sense to factor the checksum call into a private method for pure readability that contains a docstring that contains this information (cross-referencing the backend implementations)?

You would vote for still having `audbackend.core.utils.checksum()` (that has a proper docstring and just calls `audeer.md5()`), but not include it in the public API, correct?

> My understanding is that backend.put_file triggers the file upload. There is a class Interface but it is not relevant for the backend interaction.

In the interface classes, the underlying backend methods are called, e.g. `audbackend.interface.Unversioned.checksum()`:

https://github.com/audeering/audbackend/blob/1b1177de47dbe4e8eb58f19153ddc6afbd86ccae/audbackend/core/interface/unversioned.py#L38-L71

Reply:
>> You would vote for still having audbackend.core.utils.checksum() (that has a proper docstring and just calls audeer.md5()), but not include it in the public API, correct?

I am not sure - or not really. My main point is to make using the class and understanding it easy for the developer. MD5s are a kind of checksum.  I find that the conditional


```python
        if path_is_local:
            checksum = audeer.md5(path)
        else:
            checksum = self.checksum(path)
```

might trigger misunderstanding.  And I also find that putting in an intermediate layer that polishes the calling convention is a remedy but not a cure - and I don't have one either. Maybe a docstring would do?

>> In the interface classes, the underlying backend methods are called, e.g. audbackend.interface.Unversioned.checksum():

Ok yes. But what is being calculated comes from the backend, i.e. the backend devices on the  implementation - the interface is transparent.

Reply:
OK, I updated the code to use the term `md5`:

```python
        md5 = audeer.md5(path) if path_is_local else self.checksum(path)
        md5_ref = audeer.md5(path_ref) if path_ref_is_local else self.checksum(path_ref) 
```

and I extended the docstring to:

```python
        r"""Assert checksums are equal.

        Compare the MD5 sum of a file
        (``path``)
        to the MD5 sum of a reference file
        (``path_ref``).
        If check fails,
        ``path`` is removed
        and an error is raised.

        Both ``path`` and ``path_ref``
        can be local files,
        or stored on any backend.

        Args:
            path: path to a file.
                Its MD5 sum is compared
                to a reference one,
                calculated from ``path_ref``
            path_is_local: if ``True``,
                assumes ``path`` is stored on local machine
            path_ref: path to a file.
                Its MD5 sum is used as reference
            path_ref_is_local: if ``True``,
                assumes ``path_ref`` is stored on local machine

        Raises:
            InterruptedError: if the MD5 sums do not match

        """
```

Reply:
Nice.

Reply:
Thanks, I will merge then.